File: sensitivity/alt_censoring/run_alt_censoring.m
%% Alternative-censoring sensitivity
%
% Reviewer 1 asked whether CICADAS results are robust to alternative
% censoring mechanisms. The baseline DGP has treatment-dependent informative
% censoring (baseline V parameters in a0_GenerateTrialData.m). Here we
% regenerate observational data under three alternative censoring models:
%
%   (1) baseline: parmsV as in a0_GenerateTrialData.m
%   (2) MCAR:     censoring completely at random (uniform censoring rate)
%   (3) strong:   strongly outcome-dependent censoring
%   (4) none:     no censoring at all (a stress-test)
%
% For each, we run the CICADAS g-formula pipeline and report recovered
% ATE vs the RCT truth.

clear; clc; format compact;
script_dir = fileparts(mfilename('fullpath'));
repo_root  = fileparts(fileparts(script_dir));
addpath(repo_root);
addpath(genpath(fullfile(repo_root, 'CICADA_FIGURES')));

%% Shared baseline params
rng(0);
N = 1000;
th = 0.1;
parmsControl = [10, 50];
[age, sofa, C, g, parmsPD] = fcnGeneratePatientParameters(N, ...
    'TargetCMean', 3, 'TargetGMean', 4, 'CV', 0.1);
ke = 0.5;
parmsY = [-7, .3, 20, 5];
parmsL = [0.25, 1, 0.15, 0.05, 0.15, 0.03, 40];
dt = 2; t = 0:dt:168;

% Censoring regimes: parmsV = [b0 b1 b2 b3 b4 b5]
%   logit_v = Rx*(b0 + b1*cumA + b2*t^2) + (1-Rx)*(b3 + b4*cumL + b5*t^2)
regimes = struct();
regimes.baseline = [-5,  2.0, 0.1, -5,  2.0, 1.5];
regimes.MCAR     = [-6,  0.0, 0.0, -6,  0.0, 0.0];
regimes.strong   = [-3,  4.0, 0.5, -3,  4.0, 2.5];
regimes.none     = [-50, 0.0, 0.0, -50, 0.0, 0.0];
names = fieldnames(regimes);

% RCT truth (no censoring active in RCT mode)
L0_rct = fcnGenerateStochasticTrajectories(t, parmsL, N);
T1 = fcnSimulate_N_Patients(N, 1, 0.5*ones(1,N), th, C, g, ke, L0_rct, ...
                             parmsControl, parmsY, regimes.baseline, age, sofa);
[s0_true, s1_true] = fcnPlotKM(T1); close all;
ate_truth = s1_true(end) - s0_true(end);
fprintf('RCT truth ATE = %.3f\n', ate_truth);

results = struct();
for i = 1:length(names)
    nm = names{i};
    parmsV_i = regimes.(nm);
    rng(1 + i);

    L0 = fcnGenerateStochasticTrajectories(t, parmsL, N);
    treatProb = fcnBiasedAssignmentProb(age, sofa, L0(:, 1:5));
    T0 = fcnSimulate_N_Patients(N, 0, treatProb, th, C, g, ke, L0, ...
                                parmsControl, parmsY, parmsV_i, age, sofa);

    n_cens = sum(T0.V > 0);
    total_rows = height(T0);
    fprintf('\nRegime "%s": censoring rate = %.1f%%\n', nm, 100*n_cens/total_rows);

    try
        parmsY_est = fcnEstimateDeathParms(T0);
        [parmsL_est, LL, AA, age_e, sofa_e, t_e] = fcnEstimateParmsL(T0);
        [ke_est, C_est, g_est] = fcnEstimateParmsPKPD(parmsL_est, LL, AA, age_e, sofa_e, t_e);
        L0_est = fcnGenerateStochasticTrajectories(t, parmsL_est, N);
        T1_est = fcnSimulate_N_Patients(N, 1, 0.5*ones(1,N), th, C_est, g_est, ke_est, ...
                                         L0_est, parmsControl, parmsY_est, zeros(1,6), age, sofa);
        [s0_gf, s1_gf] = fcnPlotKM(T1_est); close all;
        ate = s1_gf(end) - s0_gf(end);
        bias = ate - ate_truth;
    catch ME
        fprintf('  [WARN] %s\n', ME.message);
        ate = NaN; bias = NaN;
    end
    results.(nm) = struct('ate', ate, 'bias', bias, 'cens_rate', n_cens/total_rows);
    fprintf('  ATE = %+.3f (bias %+.3f)\n', ate, bias);
end

save(fullfile(script_dir, 'alt_censoring_results.mat'), 'results','ate_truth','names','regimes');
fprintf('\nSaved results.\n');

%% Figure
figure(1); clf;
atess = zeros(1, length(names));
biases = zeros(1, length(names));
cens_rates = zeros(1, length(names));
for i = 1:length(names)
    atess(i) = 100 * results.(names{i}).ate;
    biases(i) = 100 * results.(names{i}).bias;
    cens_rates(i) = 100 * results.(names{i}).cens_rate;
end
subplot(1,2,1);
bar(atess); set(gca, 'XTickLabel', names); ylabel('G-formula ATE (pp)');
yline(100*ate_truth, 'r--', 'LineWidth', 1.5, 'Label', sprintf('RCT truth = %.1f', 100*ate_truth));
title('Recovered ATE across censoring regimes');
subplot(1,2,2);
bar(cens_rates); set(gca, 'XTickLabel', names); ylabel('Censoring rate (% of person-time)');
title('Censoring rate by regime');
set(gcf, 'Units','inches', 'Position',[1 1 12 4.5]);
set(gcf, 'PaperUnits','inches', 'PaperSize',[12 4.5], 'PaperPosition',[0 0 12 4.5]);
print(gcf, fullfile(script_dir, 'Fig_alt_censoring.pdf'), '-dpdf', '-r300');

%% Text summary
fid = fopen(fullfile(script_dir, 'alt_censoring_summary.txt'), 'w');
fprintf(fid, 'Alternative-censoring sensitivity\n');
fprintf(fid, 'RCT truth ATE: %+.3f\n\n', ate_truth);
fprintf(fid, '%-10s %-14s %-14s %-14s\n', 'regime', 'cens rate', 'ATE', 'bias');
for i = 1:length(names)
    nm = names{i};
    fprintf(fid, '%-10s %-14.3f %-+14.3f %-+14.3f\n', nm, ...
            results.(nm).cens_rate, results.(nm).ate, results.(nm).bias);
end
fclose(fid);

File: sensitivity/alt_censoring/run_alt_censoring.py
"""run_alt_censoring.py -- Python port of run_alt_censoring.m.

Alternative-censoring sensitivity. Regenerates observational data under
four censoring regimes and reports recovered ATE vs RCT truth.

Outputs (written next to this script):
    alt_censoring_results.mat
    Fig_alt_censoring.pdf
    alt_censoring_summary.txt
"""

import os
import sys
import time
import traceback

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np
from scipy.io import savemat

_SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
_REPO_ROOT = os.path.dirname(os.path.dirname(_SCRIPT_DIR))  # <repo>/cicadas
_PYTHON_DIR = os.path.join(_REPO_ROOT, "python")

sys.path.insert(0, _PYTHON_DIR)
np.random.seed(0)

from fcnGeneratePatientParameters import fcnGeneratePatientParameters  # noqa: E402
from fcn_generateStochasticTrajectories import fcnGenerateStochasticTrajectories  # noqa: E402
from fcnBiasedAssignmentProb import fcnBiasedAssignmentProb  # noqa: E402
from fcnPlotKM import fcnPlotKM  # noqa: E402
from fcnEstimateDeathParms import fcnEstimateDeathParms  # noqa: E402
from fcnEstimateParmsL import fcnEstimateParmsL  # noqa: E402
from fcnEstimateParmsPKPD import fcnEstimateParmsPKPD  # noqa: E402
from fcnSimulate_N_Patients import fcnSimulate_N_Patients  # noqa: E402


def main() -> int:
    # --- Baseline params ---
    N = 1000
    th = 0.1
    parmsControl = np.array([10.0, 50.0])
    np.random.seed(0)
    age, sofa, C, g, _ = fcnGeneratePatientParameters(
        N, "TargetCMean", 3, "TargetGMean", 4, "CV", 0.1,
    )
    ke = 0.5
    parmsY = np.array([-7.0, 0.3, 20.0, 5.0])
    parmsL = np.array([0.25, 1.0, 0.15, 0.05, 0.15, 0.03, 40.0])
    dt = 2.0
    t = np.arange(0.0, 168.0 + dt, dt)

    regimes = {
        "baseline": np.array([-5.0,  2.0, 0.1, -5.0,  2.0, 1.5]),
        "MCAR":     np.array([-6.0,  0.0, 0.0, -6.0,  0.0, 0.0]),
        "strong":   np.array([-3.0,  4.0, 0.5, -3.0,  4.0, 2.5]),
        "none":     np.array([-50.0, 0.0, 0.0, -50.0, 0.0, 0.0]),
    }
    names = list(regimes.keys())

    # --- RCT truth (no censoring in RCT mode) ---
    L0_rct = fcnGenerateStochasticTrajectories(t, parmsL, N)
    T1 = fcnSimulate_N_Patients(
        N, 1, 0.5 * np.ones(N), th, C, g, ke, L0_rct,
        parmsControl, parmsY, regimes["baseline"], age, sofa,
    )
    s0_true, s1_true, _, _ = fcnPlotKM(T1)
    ate_truth = float(s1_true[-1] - s0_true[-1])
    print(f"RCT truth ATE = {ate_truth:.3f}")

    results = {}
    t_all = time.time()
    for i, nm in enumerate(names):
        parmsV_i = regimes[nm]
        np.random.seed(1 + (i + 1))

        L0 = fcnGenerateStochasticTrajectories(t, parmsL, N)
        treatProb = fcnBiasedAssignmentProb(age, sofa, L0[:, :5])
        T0 = fcnSimulate_N_Patients(
            N, 0, treatProb, th, C, g, ke, L0,
            parmsControl, parmsY, parmsV_i, age, sofa,
        )

        n_cens = int((T0["V"] > 0).sum())
        total_rows = len(T0)
        cens_rate = n_cens / max(total_rows, 1)
        print(f"\nRegime \"{nm}\": censoring rate = {100*cens_rate:.1f}%")

        ate = float("nan"); bias = float("nan")
        try:
            parmsY_est = fcnEstimateDeathParms(T0)
            parmsL_est, LL, AA, age_e, sofa_e, t_e = fcnEstimateParmsL(T0)
            ke_est, C_est, g_est, _ = fcnEstimateParmsPKPD(
                parmsL_est, LL, AA, age_e, sofa_e, t_e,
            )
            parmsL_est = np.asarray(parmsL_est, dtype=float).reshape(-1)[:7]
            L0_est = fcnGenerateStochasticTrajectories(t, parmsL_est, N)
            T1_est = fcnSimulate_N_Patients(
                N, 1, 0.5 * np.ones(N), th, C_est, g_est, ke_est, L0_est,
                parmsControl, parmsY_est, np.zeros(6), age, sofa,
            )
            s0_gf, s1_gf, _, _ = fcnPlotKM(T1_est)
            ate = float(s1_gf[-1] - s0_gf[-1])
            bias = ate - ate_truth
        except Exception as e:  # noqa: BLE001
            print(f"  [WARN] {e}")
            traceback.print_exc()

        results[nm] = {"ate": ate, "bias": bias, "cens_rate": cens_rate}
        print(f"  ATE = {ate:+.3f} (bias {bias:+.3f})")

    print(f"\nTotal time: {(time.time() - t_all)/60:.1f} min")

    # --- Save .mat (flat keys for SciPy compatibility) ---
    mat_payload = {
        "ate_truth": ate_truth,
        "names": np.array(names, dtype=object),
    }
    for nm in names:
        mat_payload[f"ate_{nm}"]       = results[nm]["ate"]
        mat_payload[f"bias_{nm}"]      = results[nm]["bias"]
        mat_payload[f"cens_rate_{nm}"] = results[nm]["cens_rate"]
        mat_payload[f"parmsV_{nm}"]    = regimes[nm]
    out_file = os.path.join(_SCRIPT_DIR, "alt_censoring_results.mat")
    savemat(out_file, mat_payload, do_compression=True)
    print(f"Saved {out_file}")

    # --- Figure ---
    atess = np.array([100 * results[nm]["ate"]       for nm in names])
    cens_rates = np.array([100 * results[nm]["cens_rate"] for nm in names])

    fig, axes = plt.subplots(1, 2, figsize=(12, 4.5))
    axes[0].bar(names, atess)
    axes[0].axhline(100 * ate_truth, color="red", linestyle="--", linewidth=1.5,
                    label=f"RCT truth = {100*ate_truth:.1f}")
    axes[0].set_ylabel("G-formula ATE (pp)")
    axes[0].set_title("Recovered ATE across censoring regimes")
    axes[0].legend()

    axes[1].bar(names, cens_rates)
    axes[1].set_ylabel("Censoring rate (% of person-time)")
    axes[1].set_title("Censoring rate by regime")

    fig_path = os.path.join(_SCRIPT_DIR, "Fig_alt_censoring.pdf")
    plt.savefig(fig_path, bbox_inches="tight")
    plt.close(fig)
    print(f"Saved figure to {fig_path}")

    # --- Text summary ---
    txt_path = os.path.join(_SCRIPT_DIR, "alt_censoring_summary.txt")
    with open(txt_path, "w") as fid:
        fid.write("Alternative-censoring sensitivity\n")
        fid.write(f"RCT truth ATE: {ate_truth:+.3f}\n\n")
        fid.write(f"{'regime':<10} {'cens rate':<14} {'ATE':<14} {'bias':<14}\n")
        for nm in names:
            fid.write(
                f"{nm:<10} {results[nm]['cens_rate']:<14.3f} "
                f"{results[nm]['ate']:<+14.3f} {results[nm]['bias']:<+14.3f}\n"
            )
    print(f"Saved text summary to {txt_path}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
